lib_xua/src/core/endpoint0/xua_ep0_msos_descriptors.h

```diff
@@ -0,0 +1,154 @@
+
+
+typedef enum
+{
+  MS_OS_20_SET_HEADER_DESCRIPTOR       = 0x00,
+  MS_OS_20_SUBSET_HEADER_CONFIGURATION = 0x01,
+  MS_OS_20_SUBSET_HEADER_FUNCTION      = 0x02,
+  MS_OS_20_FEATURE_COMPATBLE_ID        = 0x03,
+  MS_OS_20_FEATURE_REG_PROPERTY        = 0x04,
+  MS_OS_20_FEATURE_MIN_RESUME_TIME     = 0x05,
+  MS_OS_20_FEATURE_MODEL_ID            = 0x06,
+  MS_OS_20_FEATURE_CCGP_DEVICE         = 0x07,
+  MS_OS_20_FEATURE_VENDOR_REVISION     = 0x08
+} microsoft_os_20_type_t;
+
+///////////////////////////////////////////////////////////////////////////////////////////////
+#if XUA_ENABLE_BOS_DESC
+#define BOS_TOTAL_LEN      (USB_BOS_DESC_LEN + USB_BOS_MICROSOFT_OS_DESC_LEN)
+
+unsigned char const desc_bos_runtime[] =
+{
+  // total length, number of device caps
+  USB_BOS_DESCRIPTOR(BOS_TOTAL_LEN, 1),
+
+  // Microsoft OS 2.0 descriptor
+  USB_BOS_MS_OS_20_DESCRIPTOR(MS_OS_20_DESC_LEN_RUNTIME, REQUEST_GET_MS_DESCRIPTOR)
+};
+
+unsigned char const desc_bos_dfu[] =
+{
+  // total length, number of device caps
+  USB_BOS_DESCRIPTOR(BOS_TOTAL_LEN, 1),
+
+  // Microsoft OS 2.0 descriptor
+  USB_BOS_MS_OS_20_DESCRIPTOR(MS_OS_20_DESC_LEN_DFU, REQUEST_GET_MS_DESCRIPTOR)
+};
+
+
+uint8_t desc_ms_os_20_runtime[] =
+{
+  // Set header: length, type, windows version, total length
+  U16_TO_U8S_LE(0x000A), U16_TO_U8S_LE(MS_OS_20_SET_HEADER_DESCRIPTOR), U32_TO_U8S_LE(0x06030000), U16_TO_U8S_LE(MS_OS_20_DESC_LEN_RUNTIME),
+
+  // Configuration subset header: length, type, configuration index, reserved, configuration total length
+  U16_TO_U8S_LE(0x0008), U16_TO_U8S_LE(MS_OS_20_SUBSET_HEADER_CONFIGURATION), 0, 0, U16_TO_U8S_LE(MS_OS_20_DESC_LEN_RUNTIME-0x0A),
+
+  // Function Subset header: length, type, first interface, reserved, subset length
+  U16_TO_U8S_LE(0x0008), U16_TO_U8S_LE(MS_OS_20_SUBSET_HEADER_FUNCTION), INTERFACE_NUMBER_DFU, 0, U16_TO_U8S_LE(MS_OS_20_DESC_LEN_RUNTIME-0x0A-0x08),
+
+  // MS OS 2.0 Compatible ID descriptor: length, type, compatible ID, sub compatible ID
+  U16_TO_U8S_LE(0x0014), U16_TO_U8S_LE(MS_OS_20_FEATURE_COMPATBLE_ID), 'W', 'I', 'N', 'U', 'S', 'B', 0x00, 0x00,
+  0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, // sub-compatible
+
+  // MS OS 2.0 Registry property descriptor: length, type
+  U16_TO_U8S_LE(MS_OS_20_DESC_LEN_RUNTIME-0x0A-0x08-0x08-0x14), U16_TO_U8S_LE(MS_OS_20_FEATURE_REG_PROPERTY),
+  U16_TO_U8S_LE(0x0007), U16_TO_U8S_LE(0x002A), // wPropertyDataType, wPropertyNameLength and PropertyName "DeviceInterfaceGUIDs\0" in UTF-16
+  'D', 0x00, 'e', 0x00, 'v', 0x00, 'i', 0x00, 'c', 0x00, 'e', 0x00, 'I', 0x00, 'n', 0x00, 't', 0x00, 'e', 0x00,
+  'r', 0x00, 'f', 0x00, 'a', 0x00, 'c', 0x00, 'e', 0x00, 'G', 0x00, 'U', 0x00, 'I', 0x00, 'D', 0x00, 's', 0x00, 0x00, 0x00,
+  U16_TO_U8S_LE(0x0050), // wPropertyDataLength
+  // [DriverInterface] bPropertyData
+  // InterfaceGUID = {89C14132-D389-4FF7-944E-2E33379BB59D}
+  '{', 0x00, '8', 0x00, '9', 0x00, 'C', 0x00, '1', 0x00, '4', 0x00, '1', 0x00, '3', 0x00, '2', 0x00, '-', 0x00,
```

Comment:
Going forwards we're going to ideally need a method for the customer to easily update the GUID. A define or a callback maybe.

Reply:
I've added the interface GUID as a #define in xua_conf_default.h that the user can override by defining their own in the application's xua_conf.h